### PR TITLE
BACKLOG-20594: not add alternate or canonical if they exist already

### PR DESCRIPTION
**Pull request by jahia-carp**
https://jira.jahia.org/browse/BACKLOG-20594

### CODE REVIEW

File: test-module/pom.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.jahia.modules</groupId>
+        <artifactId>site-settings-seo-parent</artifactId>
+        <version>3.6.0-SNAPSHOT</version>
+    </parent>
+    <groupId>org.jahia.test</groupId>
+    <artifactId>site-settings-seo-test-module</artifactId>
+    <name>Test module for site settings seo</name>
+    <version>1.0.0-SNAPSHOT</version>
```
Comment (jsinovassin) [inline test-module/pom.xml:60]: It should be better ta have the same version than the parent for this module

File: test-module/pom.xml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    <packaging>bundle</packaging>
+    <description>This is the custom profile module (templateSiteSettingsSEO) for running on a Jahia server.</description>
+
+    <scm>
```
Comment (jsinovassin) [inline test-module/pom.xml:64]: This part can be removed

File: test-module/src/main/resources/META-INF/rules.drl.disabled [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.jahia.test.sitesettingsseotestmodule.rules
```
Comment (jsinovassin) [inline test-module/src/main/resources/META-INF/rules.drl.disabled:1]: You can delete this file as it is not used

File: test-module/src/main/resources/jnt_template/html/template.with-seo-links.jsp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    <link rel="canonical" href="http://localhost/canonical/home/en">
+	<link rel="alternate" hreflang="en" href="http://localhost/alternate/home/en">
+    <link rel="alternate" hreflang="fr" href="http://localhost/alternate/home/fr">
+</head>
+
+<body>
+
+<h2 data-sel="with-seo-link">TEMPLATE SET WITH SEO LINKS</h2>
+
+<div class="bodywrapper"><!--start bodywrapper-->
```
Comment (jsinovassin) [inline test-module/src/main/resources/jnt_template/html/template.with-seo-links.jsp:31]: Not a big problem but I think comments like this does not bring much information in the code

**Comment by jsinovassin** [comment]
site-settings-seo-test-module.properties file can be removed as well

File: test-module/src/site/apt/index.apt [hunk 5]
```diff
@@ -0,0 +1 @@
+site-settings-seo-test-module
```
Comment (jsinovassin) [inline test-module/src/site/apt/index.apt:1]: Do we know if this file is used somewhere?

File: test-module/src/site/site.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jsinovassin) [inline test-module/src/site/site.xml:1]: Do we need this file?

File: tests/cypress/e2e/basic.test.cy.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import { publishAndWaitJobEnding } from '../utils/publishAndWaitJobEnding'
+
+describe("Basic tests of the module's seo filter", () => {
+    const siteKey = 'headLinksTest'
+    const sitePath = '/sites/' + siteKey
+    const homePath = sitePath + '/home'
+    const templateSet = 'site-settings-seo-test-module'
+
+    const createPage = (parent, name, template = undefined) => {
```
Comment (jsinovassin) [inline tests/cypress/e2e/basic.test.cy.ts:9]: Is it necessary to set the default value to undefined for template?

File: tests/cypress/fixtures/graphql/jcrPublicationStatus.graphql [hunk 8]
```diff
@@ -0,0 +1,10 @@
+query ($path: String!) {
+    jcr {
+        nodeByPath(path: $path) {
+            uuid
+            aggregatedPublicationInfo(language: "en", subNodes: true) {
```
Comment (jsinovassin) [inline tests/cypress/fixtures/graphql/jcrPublicationStatus.graphql:5]: Not mandatory, but it could be nice to have language as a parameter

**Review by jsinovassin — CHANGES_REQUESTED** [review]

File: test-module/pom.xml (hunk 1 above)
Comment (jahia-carp) [reply test-module/pom.xml]: Fixed

File: test-module/pom.xml (hunk 2 above)
Comment (jahia-carp) [reply test-module/pom.xml]: Done

File: test-module/src/main/resources/META-INF/rules.drl.disabled (hunk 3 above)
Comment (jahia-carp) [reply test-module/src/main/resources/META-INF/rules.drl.disabled]: Done

File: test-module/src/main/resources/META-INF/site-settings-seo-test-module.tld.disabled [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
[the remark this answers is not in the document]
Comment (jahia-carp) [reply test-module/src/main/resources/META-INF/site-settings-seo-test-module.tld.disabled]: Done

File: test-module/src/main/resources/jnt_template/html/template.with-seo-links.jsp (hunk 4 above)
Comment (jahia-carp) [reply test-module/src/main/resources/jnt_template/html/template.with-seo-links.jsp]: Done

File: test-module/src/site/apt/index.apt (hunk 5 above)
Comment (jahia-carp) [reply test-module/src/site/apt/index.apt]: Removed

File: test-module/src/site/site.xml (hunk 6 above)
Comment (jahia-carp) [reply test-module/src/site/site.xml]: Removed

File: tests/cypress/e2e/basic.test.cy.ts (hunk 7 above)
Comment (jahia-carp) [reply tests/cypress/e2e/basic.test.cy.ts]: Fixed not really necessary for this test

**Review by jkevan — CHANGES_REQUESTED** [review]
I'm not sure that the algorithm implement is correct.
I understand that we wanted:
- do not calculate **alternates** in case there is already 1+ **alternate** link tag
- do not calculate **canonical** in case there is already 1+ **canonical** link tag

But here it seem's that:
- in case there is a **canonical** tag, but no **alternates** tag, then **alternates** tags wont be calculated. (and I think we should calculate the **alternates** because there is no **alternates** tags, but should not calculate the **canonical** only.)

So the detection should be smarter and make the difference between canonical and alternates presence.
To be confirmed with PM, but it seem's that's what we agreed on during the 3 amigos session.

File: tests/cypress/fixtures/graphql/jcrPublicationStatus.graphql (hunk 8 above)
Comment (jahia-carp) [reply tests/cypress/fixtures/graphql/jcrPublicationStatus.graphql]: Done

**Review by jsinovassin — APPROVED** [review]

**Review by jkevan — APPROVED** [review]

**Pull request merged**